Détection et implutation

In [3]:
import pandas as pd
import numpy as np
data ={
 'patient_ID':[101,101,103,104,105],
 'age':[25 , 30 , np . nan , 45 , 22] ,
 'tension':[12 , np . nan , 14 , 15 , np . nan ],
 'Fumeur':['Oui', 'Non' , 'Non', np . nan , 'Oui' ]
}
df=pd.DataFrame(data)

# 1.Détection global

In [4]:
print(df.isnull().sum())

patient_ID    0
age           1
tension       2
Fumeur        1
dtype: int64


# 2.imputation de l'age par le moyenne

In [5]:
df['age']=df['age'].fillna(df['age'].mean())

In [6]:
df['tension']=df['tension'].fillna(df['tension'].median)

In [7]:
print(df.isnull().sum())

patient_ID    0
age           0
tension       0
Fumeur        1
dtype: int64


utilisation de mode pour complet le vide

In [8]:
df['Fumeur']=df['Fumeur'].fillna(df['Fumeur'].mode)

In [9]:
print(df.isnull().sum())

patient_ID    0
age           0
tension       0
Fumeur        0
dtype: int64


# 3.pourquoi parfois risque de supprimer systématiquement toutes les lignes contenant des NaNs

car les donnees restantes ne sont plus representatives (ne representent plus la realite)

# exercice2

Detection par IQR

# SIMULATION DE PRIX (EN Millions MRU)

In [10]:
prix= pd.Series([1.2 , 1.5 , 1.1 , 120.0 , 1.4 , 1.3 , 0.9 , 85.0 , 1.2])
Q1=prix.quantile(0.25)
Q3=prix.quantile(0.75)
IQR=Q3-Q1
# definition de bornes
borne_inf = Q1 - 1.5 * IQR
borne_sup = Q3 + 1.5 * IQR
outliers = prix [( prix < borne_inf ) | ( prix > borne_sup ) ]
print (f" Valeurs ␣ aberrantes ␣ detectees ␣:␣{ outliers . values }")


 Valeurs ␣ aberrantes ␣ detectees ␣:␣[120.  85.]


CALCULER Z-SCORE

X=120,
μ=23,73 #MOYENNE,
σ=43,
Z=(x −μ)/σ),
DONC Z=2,23 INTERPRETATION:120 TRES PETIT OUTLIER


In [11]:
prix=prix.clip(lower=borne_inf,upper=borne_sup)

3. Quel graphique (vu en cours) permet de visualiser instantanément les out-
liers ?

cest Box Plot(Diagramme en boite)

Exercice 3 :Normalisation et Encodage

In [13]:
from sklearn.preprocessing import MinMaxScaler,LabelEncoder

In [22]:
# Donnees de test
df_iris = pd.DataFrame ({
    'Taille_Cm': [150 , 180 , 160 , 200] ,
    'Couleur':['Rouge','Bleu','Vert','Rouge']
})

In [21]:
# 1. Min - Max Scaling
scaler = MinMaxScaler ()
df_iris [' Taille_Scaled '] = scaler.fit_transform ( df_iris [[ 'Taille_Cm']])

# 2. Label Encoding pour la couleur
encoder = LabelEncoder ()
df_iris ['Couleur_Enc'] = encoder . fit_transform ( df_iris ['Couleur'])

1. Expliquez la différence entre Label Encoding et One-Hot Encoding. Laquelle
utiliseriez-vous pour une colonne "Nationalité" ? Justifiez.

1.label encoding:Transforme chaque categorie en nombre entier

2.one-hot-encoding:Transforme chaque categorie en colonne binaire(0/1)

In [16]:
from sklearn.preprocessing import StandardScaler

applique sur la taille_cm

In [23]:
scaler=StandardScaler()
df_iris['Taille_scaled']=scaler.fit_transform(df_iris[['Taille_Cm']])

verifier la moyenne

In [25]:
df_iris['Taille_scaled'].mean()

np.float64(0.0)

3. Pourquoi la normalisation est-elle cruciale avant d’utiliser l’algorithme des
K-Plus Proches Voisins (KNN) ?

1.équilibrer l'influence des variables:chaque feature contribue équitablement

2.Amèliorer la distance:les voisins choisis deviennent réellement proches

3.Amèliorer la performence

EXERCICE 4

In [31]:
import pandas as pd
from sklearn . impute import SimpleImputer
from sklearn . preprocessing import StandardScaler , OneHotEncoder

In [32]:
# 1. Chargement du dataset Titanic
df_titanic = pd.read_csv('titanic.csv')

In [37]:
# 2. Traitement des valeurs manquantes ( Imputation par strategie )
 # Imputation par la constante " Inconnu " pour les cabines
df_titanic['Cabin']= df_titanic['Cabin'].fillna('Inconnu')
# 3. Traitement du dataset UCI ( Cancer )
# Note : Souvent les valeurs manquantes y sont notees ’?’
df_cancer = pd.read_csv ('risk_factors_cervical_cancer.csv',na_values ='?')
# Suppression des colonnes ayant plus de 50% de donnees manquantes
seuil = len ( df_cancer ) * 0.5
df_cancer = df_cancer . dropna ( thresh = seuil,axis =1)

In [39]:
df_cancer.head(15)

,Age,Number of sexual partners,First sexual intercourse,Num of pregnancies,Smokes (packs/year),Hormonal Contraceptives (years)
0,18,4,15,1.0,0.0,0.00
1,15,1,14,1.0,0.0,0.00
2,52,5,16,4.0,37.0,3.00
3,46,3,21,4.0,0.0,15.00
4,42,3,23,2.0,0.0,0.00
5,51,3,17,6.0,3.4,0.00
6,26,1,26,3.0,0.0,2.00
7,45,1,20,5.0,0.0,0.00
8,44,3,26,4.0,0.0,2.00
9,27,1,17,3.0,0.0,8.00


letape de feature Engineering(Titanic-Age)

In [41]:
df_titanic['Age']=df_titanic.groupby('Pclass')['Age'].transform(lambda x:x.fillna(x.mean()))

encoding titanic

In [44]:
df_titanic['Sex']=df_titanic['Sex'].map({'male':0 ,'female':1})

Embarked One-Hot Encoding

In [45]:
df_titanic=pd.get_dummies(df_titanic,columns=['Embarked'])

nettoyage (Cancer DataSet)

In [46]:
df_cancer=df_cancer.apply(pd.to_numeric,errors='coerce')

Stanrdisation

In [47]:
scaler=StandardScaler()
df_cancer_scaled=scaler.fit_transform(df_cancer)

verification de doublons

In [48]:
df_cancer.duplicated().sum()

np.int64(128)

on le supprime

In [49]:
df_cancer.drop_duplicates()

,Age,Number of sexual partners,First sexual intercourse,Num of pregnancies,Smokes (packs/year),Hormonal Contraceptives (years)
0,27.174672,4,15,1.0,0.0,0.00
1,27.174672,1,14,1.0,0.0,0.00
2,27.174672,5,16,4.0,37.0,3.00
3,27.174672,3,21,4.0,0.0,15.00
4,27.174672,3,23,2.0,0.0,0.00
...,...,...,...,...,...,...
682,27.174672,3,18,0.0,0.0,0.00
683,27.174672,2,19,1.0,0.0,8.00
684,27.174672,2,17,0.0,0.0,0.08
685,27.174672,2,24,2.0,0.0,0.08


#EXERCICE 5

1.mystere de la moyenne
reponse : les salaires imputes seront artificiellement gonflés(effet outier)
pourquoi:un seul salaire enormé tire la moyenne vers le haut.donc si tu remplace les valeurs manquantes par la moyenne,tu injectes une valeur trop élevée pour les autres.
2.l'agent double
éTAPE DE NETTOYAGE:
normalisation   /STANDARISATION DES TEXTES
3.EST UN OUTLIER,à PROUVER AVEC IQR ou z-score
4.parce que les unités et les échelles sont differentes,donc il faut scaling
5. 